### ISSUE TITLE
Invalid uri rewrite (missing last part of uri_capture) caused by other Ingresses

### ISSUE THREAD

**Issue by mrfijas** [Kong/kubernetes-ingress-controller#958]
### Summary

I'm using `request-transformer` as a way to rewrite paths of some Ingresses. Everything works as expected in a single Ingress/KongPlugin setup, but starts misbehaving when I add other Ingresses.  
My main Ingress + plugin definition:
```yaml
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: users
  annotations:
    konghq.com/plugins: users-rewrite
spec:
  ingressClassName: kong
  rules:
    - http:
        paths:
          - path: /api/users/v2(.*)
            pathType: Prefix
            backend:
              service:
                name: debug-backend
                port:
                  number: 80

---
apiVersion: configuration.konghq.com/v1
kind: KongPlugin
metadata:
  name: users-rewrite
config:
  replace:
    uri: /api/v2$(uri_captures[1])
plugin: request-transformer
```
with that configuration request to `/api/users/v2/some/other/parts` is proxied to a backend with path: `/api/v2/some/other/parts`, which is expected.

Adding other unrelated Ingresses results in call to backend with `/api/v2/some/other` (without last part of url: `/parts`).

What is also interesting problem didn't occur on setup with cluster in `1.18.2` (via `kind 0.8.1`)

Other weird thing is that bug happens only when applying 2 additional Ingresses.. with one it works just fine..

**Kong Ingress controller version**
kong-1.11.0

**Kong or Kong Enterprise version**
Kong 2.1.4

### Kubernetes version
```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.0", GitCommit:"e19964183377d0ec2052d1f1fa930c4d7575bd50", GitTreeState:"clean", BuildDate:"2020-08-26T14:30:33Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.1", GitCommit:"206bcadf021e76c27513500ca24182692aabd17e", GitTreeState:"clean", BuildDate:"2020-09-14T07:30:52Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
```

### Environment

- **Cloud provider or hardware configuration**:
  standard `kind create cluster`
- **OS** (e.g. from /etc/os-release):
  Ubuntu 20.10
- **Kernel** (e.g. `uname -a`):
  5.8.0-29-generic #31-Ubuntu SMP Fri Nov 6 12:37:59 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
- **Other**:
	```
	> kind version
	kind v0.9.0 go1.13.8 linux/amd64
	
	> helm version
	version.BuildInfo{Version:"v3.3.0", GitCommit:"8a4aeec08d67a7b84472007529e8097ec3742105", GitTreeState:"dirty", GoVersion:"go1.14.7"}
	```

### What happened

request is proxied to a backend with missing last part of captured uri group

### Expected behavior

request is proxied to a proper path 

### Steps To Reproduce

1. create cluster
	```kind create cluster```
2. install kong
	```helm install kong/kong --generate-name --set ingressController.installCRDs=false```
3. apply resources (with debug-y backend to easily see url and headers sent by proxy)
	```
	echo "
	---
	apiVersion: apps/v1
	kind: Deployment
	metadata:
	  name: debug-backend
	  labels:
	    service: debug-backend
	spec:
	  replicas: 1
	  selector:
	    matchLabels:
	      service: debug-backend
	  template:
	    metadata:
	      labels:
	        service: debug-backend
	    spec:
	      containers:
	        - name: debug-backend
	          image: brndnmtthws/nginx-echo-headers
	
	---
	apiVersion: v1
	kind: Service
	metadata:
	  name: debug-backend
	spec:
	  ports:
	  - port: 80
	    targetPort: 8080
	  selector:
	    service: debug-backend
	  sessionAffinity: None
	  type: ClusterIP
	
	---
	apiVersion: networking.k8s.io/v1
	kind: Ingress
	metadata:
	  name: users
	  annotations:
	    konghq.com/plugins: users-rewrite
	spec:
	  ingressClassName: kong
	  rules:
	    - http:
	        paths:
	          - path: /api/users/v2(.*)
	            pathType: Prefix
	            backend:
	              service:
	                name: debug-backend
	                port:
	                  number: 80
	
	---
	apiVersion: configuration.konghq.com/v1
	kind: KongPlugin
	metadata:
	  name: users-rewrite
	config:
	  replace:
	    uri: /api/v2\$(uri_captures[1])
	plugin: request-transformer
	
	---
	apiVersion: networking.k8s.io/v1
	kind: Ingress
	metadata:
	  name: apps-v1
	spec:
	  ingressClassName: kong
	  rules:
	    - http:
	        paths:
	          - path: /api/apps/v1
	            pathType: Prefix
	            backend:
	              service:
	                name: debug-backend
	                port:
	                  number: 80
	
	---
	apiVersion: networking.k8s.io/v1
	kind: Ingress
	metadata:
	  name: apps-v2
	spec:
	  ingressClassName: kong
	  rules:
	    - http:
	        paths:
	          - path: /api/apps/v2
	            pathType: Prefix
	            backend:
	              service:
	                name: debug-backend
	                port:
	                  number: 80
	" | kubectl apply -f -
	```
4. expose kong-proxy
	```
	 kubectl port-forward svc/$(helm list -q)-kong-proxy 8000:80
	```
5. make a request
	```
	curl --no-progress-bar localhost:8000/api/users/v2/some/other/parts 2>&1 | grep GET
	```
	it should read `GET /api/v2/some/other/parts HTTP/1.1`, but it is missing last `/parts`

#### Bonus steps
6. remove somehow "conflicting" Ingress:
	```
	kubectl delete ingress apps-v1
	```
7. make a request
	```
	curl --no-progress-bar localhost:8000/api/users/v2/some/other/parts 2>&1 | grep GET
	```
	now it's valid `GET /api/v2/some/other/parts HTTP/1.1`

#### Bonus logs
* response from `/routes` admin API
	```json
	{
	  "next": null,
	  "data": [
	    {
	      "id": "35cd0b32-0c27-5996-8e28-af133097f424",
	      "path_handling": "v0",
	      "paths": [
	        "\/api\/apps\/v2$",
	        "\/api\/apps\/v2\/"
	      ],
	      "destinations": null,
	      "headers": null,
	      "protocols": [
	        "http",
	        "https"
	      ],
	      "created_at": 1605711217,
	      "snis": null,
	      "service": {
	        "id": "cce963fe-47d8-5666-b6e3-e07c65023806"
	      },
	      "name": "default.apps-v2.00",
	      "hosts": null,
	      "preserve_host": true,
	      "regex_priority": 200,
	      "methods": null,
	      "sources": null,
	      "updated_at": 1605711217,
	      "https_redirect_status_code": 426,
	      "tags": null,
	      "strip_path": false
	    },
	    {
	      "id": "7b30f708-be57-5292-86a0-a586c3e78b0d",
	      "path_handling": "v0",
	      "paths": [
	        "\/api\/users\/v2(.*)$",
	        "\/api\/users\/v2(.*)\/"
	      ],
	      "destinations": null,
	      "headers": null,
	      "protocols": [
	        "http",
	        "https"
	      ],
	      "created_at": 1605711217,
	      "snis": null,
	      "service": {
	        "id": "cce963fe-47d8-5666-b6e3-e07c65023806"
	      },
	      "name": "default.users.00",
	      "hosts": null,
	      "preserve_host": true,
	      "regex_priority": 200,
	      "methods": null,
	      "sources": null,
	      "updated_at": 1605711217,
	      "https_redirect_status_code": 426,
	      "tags": null,
	      "strip_path": false
	    },
	    {
	      "id": "5615f28b-b474-5775-ae49-dd1871d245ad",
	      "path_handling": "v0",
	      "paths": [
	        "\/api\/apps\/v1$",
	        "\/api\/apps\/v1\/"
	      ],
	      "destinations": null,
	      "headers": null,
	      "protocols": [
	        "http",
	        "https"
	      ],
	      "created_at": 1605711217,
	      "snis": null,
	      "service": {
	        "id": "cce963fe-47d8-5666-b6e3-e07c65023806"
	      },
	      "name": "default.apps-v1.00",
	      "hosts": null,
	      "preserve_host": true,
	      "regex_priority": 200,
	      "methods": null,
	      "sources": null,
	      "updated_at": 1605711217,
	      "https_redirect_status_code": 426,
	      "tags": null,
	      "strip_path": false
	    }
	  ]
	}
	```

* response from `/plugins` admin API
	```json
	{
	  "next": null,
	  "data": [
	    {
	      "created_at": 1605711217,
	      "config": {
	        "remove": {
	          "querystring": [],
	          "headers": [],
	          "body": []
	        },
	        "replace": {
	          "querystring": [],
	          "headers": [],
	          "body": [],
	          "uri": "\/api\/v2$(uri_captures[1])"
	        },
	        "http_method": null,
	        "add": {
	          "querystring": [],
	          "headers": [],
	          "body": []
	        },
	        "append": {
	          "querystring": [],
	          "headers": [],
	          "body": []
	        },
	        "rename": {
	          "querystring": [],
	          "headers": [],
	          "body": []
	        }
	      },
	      "id": "bfe8ccd7-82c1-5445-857f-0cf41555a4b0",
	      "service": null,
	      "enabled": true,
	      "name": "request-transformer",
	      "protocols": [
	        "http",
	        "https"
	      ],
	      "consumer": null,
	      "tags": null,
	      "route": {
	        "id": "7b30f708-be57-5292-86a0-a586c3e78b0d"
	      }
	    }
	  ]
	}
	```

* response from `/upstreams` admin API
	```json
	{
	  "next": null,
	  "data": [
	    {
	      "client_certificate": null,
	      "created_at": 1605711217,
	      "id": "2fa8d129-7554-5d9c-b91c-9b73259c42ad",
	      "algorithm": "round-robin",
	      "name": "debug-backend.default.80.svc",
	      "tags": null,
	      "hash_fallback": "none",
	      "hash_fallback_header": null,
	      "healthchecks": {
	        "active": {
	          "https_verify_certificate": true,
	          "type": "http",
	          "http_path": "\/",
	          "timeout": 1,
	          "unhealthy": {
	            "http_statuses": [
	              429,
	              404,
	              500,
	              501,
	              502,
	              503,
	              504,
	              505
	            ],
	            "tcp_failures": 0,
	            "timeouts": 0,
	            "http_failures": 0,
	            "interval": 0
	          },
	          "healthy": {
	            "http_statuses": [
	              200,
	              302
	            ],
	            "interval": 0,
	            "successes": 0
	          },
	          "https_sni": null,
	          "concurrency": 10
	        },
	        "threshold": 0,
	        "passive": {
	          "unhealthy": {
	            "http_failures": 0,
	            "http_statuses": [
	              429,
	              500,
	              503
	            ],
	            "tcp_failures": 0,
	            "timeouts": 0
	          },
	          "healthy": {
	            "successes": 0,
	            "http_statuses": [
	              200,
	              201,
	              202,
	              203,
	              204,
	              205,
	              206,
	              207,
	              208,
	              226,
	              300,
	              301,
	              302,
	              303,
	              304,
	              305,
	              306,
	              307,
	              308
	            ]
	          },
	          "type": "http"
	        }
	      },
	      "hash_on_cookie": null,
	      "hash_on": "none",
	      "hash_on_cookie_path": "\/",
	      "host_header": null,
	      "hash_on_header": null,
	      "slots": 10000
	    }
	  ]
	}
	```

**Comment by rainest** [comment]
Does adding [regex priority annotations](https://docs.konghq.com/kubernetes-ingress-controller/1.0.x/references/annotations/#konghqcomregex-priority) help? Alternately, do you see the behavior you want if you change the non-regular expression paths to ImplementationSpecific?

Under the hood, the controller actually creates Kong regular expression route paths for Prefix and Exact pathTypes, since there's not an easy way to reconcile the expectations of the Ingress spec with the standard behavior of the Kong proxy without using regular expressions. That complicates matching precedence, however, since precedence order for regular expression paths is essentially arbitrary without explicit priorities (there's some [finer nuance to how the tiebreaking actually works](https://docs.konghq.com/2.2.x/proxy/#matching-priorities), but in practice it's easier to just set priorities). As such, there are two options to sort that out:

- Add annotations indicating priority, so that your more-specific regexes are evaluated first.
- Use ImplementationSpecific pathTypes for paths that don't require regexes, which allows them to use plain Kong route paths. Plain path routes naturally take lower precedence than regex path routes in matching.

**Comment by mrfijas** [comment]
thank you for detailed explanation! I did first tests with `ImplementationSpecific` pathType and it fixes the problem (for now in simple problem-reproduction environment - I'll apply that in our real cluster soon with full ingress setup).
interestingly it works when applying to either non-regex-y paths or regex path, so:
```
- path: /api/apps/v2
  pathType: ImplementationSpecific
- path: /api/apps/v1
  pathType: ImplementationSpecific
- path: /api/users/v2(.*)
  pathType: Prefix
```
works,
```
- path: /api/apps/v2
  pathType: Prefix
- path: /api/apps/v1
  pathType: Prefix
- path: /api/users/v2(.*)
  pathType: ImplementationSpecific
```
works also.

although it seems to work right now, I'll make some additional experiments with regex priorities.

to better understand issue: is there any chance to peek at kong's internal config "computed" from these ingresses?
with `ingress-nginx` I can see given `location`s in `/etc/nginx/nginx.conf`.
in kong proxy I can see (`/kong_prefix/nginx-kong.conf`):
```
location / {
	..
    proxy_pass            $upstream_scheme://kong_upstream$upstream_uri;
}
```
so pardon the silly question - but how this `kong_upstream` is proxying requests? does it have some nginx config or is it using some internal proxy?

probably seeing resolved config could help me understand the issue - because now I'm really confused why precedence of matching is resulting in "almost" working rewrites (it's not that requests aren't matched, or totally improperly rewritten - it's just loosing last part of capturing group)

**Comment by rainest** [comment]
> to better understand issue: is there any chance to peek at kong's internal config "computed" from these ingresses?

The `/routes` admin API output is effectively it. For the Prefix paths, you'll see that those are transformed into two separate paths:

```
       "paths": [
         "\/api\/apps\/v2$",
         "\/api\/apps\/v2\/"
       ],
```

The actual resolution happens [in Lua code](https://github.com/Kong/kong/blob/master/kong/router.lua), though it's not particularly easy to follow (it's optimized more for performance than readability, and while there are additional internal computed data structures, there's no way to view them). It chooses a route based on the configuration in `/routes`, determines where it needs to go based on the attached service, and populates the variables you see in the `proxy_pass` directive.

Sending `Kong-Debug: 1` in your request headers will add response headers indicating the matched route and service IDs, and I'd be curious if it's actually matching one of the others here when everything is a Prefix path--on further review I don't think there should be any chance the expressions are overlapping, since the static components of the expression, `/api/apps` and `/api/users`, are mutually exclusive--accidentally creating overlapping regexes is a common problem, but it shouldn't actually happen here.

**Comment by mrfijas** [comment]
ooh, I've tried earlier with `X-Kong-Debug` instead of `Kong-Debug` :facepalm: 

looking at `Kong-Route-Id` it's always the same route, with prefix `/api/users/v2` - it would be really weird otherwise, as you stated.. they are mutually exclusive.

tbh I cannot find what really causes the problem here.. routes in `/routes` looks the alright (for example it isn't the case of flipping order ` "\/api\/users\/v2(.*)$"` and `"\/api\/users\/v2(.*)\/"` or anything like this).

interestingly adding these two additional paths (`/api/apps/v1` and `/api/apps/v2`) to the same ingress doesn't break `users-rewrite`:
```yaml
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: users
  annotations:
    konghq.com/plugins: users-rewrite
spec:
  ingressClassName: kong
  rules:
    - http:
        paths:
          - path: /api/users/v2(.*)
            pathType: Prefix
            backend:
              service:
                name: debug-backend
                port:
                  number: 80
          - path: /api/apps/v2
            pathType: Prefix
            backend:
              service:
                name: debug-backend
                port:
                  number: 80
          - path: /api/apps/v1
            pathType: Prefix
            backend:
              service:
                name: debug-backend
                port:
                  number: 80
```
works fine, but adding these two paths into separate ingress causes `users-rewrite` to loose last part of url
```yaml
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: apps-v2
spec:
  ingressClassName: kong
  rules:
    - http:
        paths:
          - path: /api/apps/v2
            pathType: Prefix
            backend:
              service:
                name: debug-backend
                port:
                  number: 80
          - path: /api/apps/v1
            pathType: Prefix
            backend:
              service:
                name: debug-backend
                port:
                  number: 80
```


thank you for all the time, insight and quick solution with `pathType: ImplementationSpecific` - it does make sense not to duplicate these paths in case of `(.*)` capture group at the end.

I think we can close this ticket for now, and maybe if I have more time to investigate it further I'll re-open?

**Comment by rainest** [comment]
We should likely recommend (or even enforce) ImplementationSpecific for any regex path: you're already fully in control of what the expression matches (if you want an exact match, include a terminating `$`; if not, omit it), so there's not much reason to use the other types. What actually occurs here probably does warrant further investigation, but it shouldn't matter if you're providing a non-regex route.

After some testing, I'm really not quite sure what's going on here. Initially, I thought that the evaluation order would reflect what's shown in the admin API, but we already create the `$`-terminated regular expression (which would include `parts` in the capture group) first.

Inside the router code, we build an array of all regex paths ordered by their route's priority (or creation time as a tiebreaker) and walk through that when trying to find a match. I added some debug code to dump that array right before we [start to try matches](https://github.com/Kong/kong/blob/2.2.0/kong/router.lua#L1442) and observed something quite confusing: the order does not match the order shown in the admin API, does not appear to be obviously based on some other criteria (e.g. alphabetical), and does appear to exhibit the behavior you describe: the order changes based on other routes.

The order does appear somehow deterministic based on the complete route set: if I add a route and then remove it, the effective order of some other existing route's expressions will revert to whatever it was before adding the new route.

I don't see anything that appears to explain this, as all intentional sorting is based on the route properties. Within the route, we [simply loop over the paths and add them to the table of all regex URIs](https://github.com/Kong/kong/blob/2.2.0/kong/router.lua#L625-L633), which shouldn't deviate from the admin API order and shouldn't take other routes' configuration into account, but seemingly does.

**Comment by mrfijas** [comment]
thank you, for digging deeper and also for https://github.com/Kong/docs.konghq.com/pull/2477 which definitely helps avoiding my mistake with `Prefix` and regex paths.

although I'm not familiar with internals of kong and generally lua, let me know if I can be of any help.

**Comment by rainest** [comment]
Nah--we shouldn't need anything more from your end. I wanted to write that down for future reference, as the core proxy behavior is weird and not something I fully understand. Best course of action should be to just avoid potential conflicts between user-provided regular expressions and our transformations, so I've added a warning to call that out.

Closing this per the earlier comment now that it's documented; thanks for helping us find this problem case!
